File: SolvedProblems/CPfor100Days/Python/tree.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.height = 0

class Tree:
    def __init__(self):
        self.root = None

    def insert(self, val):
        if self.root is None:
            self.root = Node(val)
        else:
            self.populate(self.root, val)

    def populate(self, node, val):
        if val < node.val:
            if node.left is None:
                node.left = Node(val)
                return

            self.populate(node.left, val)
        
        if val > node.val:
            if node.right is None:
                node.right = Node(val)
                return

            self.populate(node.right, val)
        
        node.height = max(self.height(node.left), self.height(node.right)) + 1
    
    def height(self, node):
        if node is None:
            return -1

        return node.height
    
    def display(self):
        self.pretty_display(self.root, 0)
        
    def pretty_display(self, node, level):
        if node is None:
            return

        self.pretty_display(node.left, level + 1)
        if level != 0:
            for i in range(level):
                print(end="\t")
            print("|---->", node.val)
        else:
            print(node.val)
        self.pretty_display(node.right, level + 1)
    
    

root = Tree()
root.insert(15)
root.insert(7)
root.insert(21)
root.insert(4)
root.insert(10)
root.insert(17)
root.insert(24)

root.display()

File: SolvedProblems/Trees/Problems/Tree.java
package Trees.Problems;

public class Tree {
    Node root;

    public void populate(int value) {
        if (root == null) {
            root = new Node(value);
        } else {
            populate(root, value);
        }
    }

    private void populate(Node node, int value) {

        if (value < node.value) {
            if (node.left == null) {
                node.left = new Node(value);
                return;
            }
            // Travel at the left side
            populate(node.left, value);
            return;
        }

        if (value > node.value) {
            if (node.right == null) {
                node.right = new Node(value);
                return;
            }
            // Travel at the right side
            populate(node.right, value);
            return;
        }
        node.height = Math.max(height(node.left), height(node.right)) + 1;

    }

    public void insert(int[] arr) {
        insert(arr, 0, arr.length - 1);
    }

    private void insert(int[] arr, int start, int end) {
        if (start > end) {
            return;
        }

        int mid = (start + end) / 2;
        populate(arr[mid]);
        insert(arr, start, mid);
        insert(arr, mid + 1, end);
    }

    private int height(Node node) {
        if (node == null) {
            return -1;
        }
        return node.height;
    }

    public void display() {
        prettyDisplay(root, 0);
    }

    private void prettyDisplay(Node node, int level) {
        if (node == null) {
            return;
        }

        prettyDisplay(node.right, level + 1);
        if (level != 0) {
            for (int i = 0; i < level; i++) {
                System.out.print("|\t");
            }
            System.out.println("|---->" + node.value);
        } else {
            System.out.println(node.value);
        }
        prettyDisplay(node.left, level + 1);

    }

}
